pkg/roll/run_update.go

```diff
@@ -360,84 +372,119 @@ func (u *update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNod
 			// don't check health if the update isn't even done there yet
 			continue
 		}
-		if hres, ok := checks[node]; ok && hres.Status == health.Passing {
-			ret.Healthy++
+		if hres, ok := checks[node]; ok {
+			if hres.Status == health.Passing {
+				ret.Healthy++
+			} else if hres.Status != health.Unknown {
+				ret.Unhealthy++
+			}
 		}
 	}
+
+	// We set unknown here to deal with the case of nodes not being in CurrentPods()
+	ret.Unknown = ret.Desired - ret.Healthy - ret.Unhealthy
+
 	return ret, err
 }
 
-func (u *update) shouldRollAfterDelay(newFields rcf.RC) (int, error) {
+func (u *update) shouldRollAfterDelay(newFields rcf.RC) (int, int, error) {
 	// Check health again following the roll delay. If things have gotten
 	// worse since we last looked, or there is an error, we break this iteration.
 	checks, err := u.hcheck.Service(newFields.Manifest.ID().String())
 	if err != nil {
-		return 0, util.Errorf("Could not retrieve health following delay: %v", err)
+		return 0, 0, util.Errorf("Could not retrieve health following delay: %v", err)
 	}
 
 	afterDelayNew, err := u.countHealthy(u.NewRC, checks)
 	if err != nil {
-		return 0, util.Errorf("Could not determine new service health: %v", err)
+		return 0, 0, util.Errorf("Could not determine new service health: %v", err)
 	}
 
 	afterDelayOld, err := u.countHealthy(u.OldRC, checks)
 	if err != nil {
-		return 0, util.Errorf("Could not determine old service health: %v", err)
+		return 0, 0, util.Errorf("Could not determine old service health: %v", err)
 	}
 
-	afterDelayNext := rollAlgorithm(afterDelayOld.Healthy, afterDelayNew.Healthy, u.DesiredReplicas, u.MinimumReplicas)
+	// We conservatively treat Unknown nodes as healthy on the old side.
+	afterDelayRemove, afterDelayNext := rollAlgorithm(
+		afterDelayOld.Healthy+afterDelayOld.Unknown, afterDelayNew.Healthy,
+		afterDelayOld.Desired+afterDelayNew.Desired, u.DesiredReplicas,
```

Comment:
Should these summary objects be parameters to `rollAlgorithm`? I ask because the decision to sum the health and unknown is copied to multiple locations in this file, and maybe we want tests to verify that it does the right thing.